util: Add MR monitor

shefty opened the pull request:
The MR monitor is a base implementation for memory management
notifications.  It will be used by a memory registration cache
in order to receive notifications of memory regions that are
modified.  For example a monitor region may be freed by the
application or the underlying pages modified.  The MR monitor
will report such changes to the user.

As a base implementation, the MR monitor doesn't do anything
other than define the function prototypes for what will become
the 'real' monitors.

Signed-off-by: Dmitry Gladkov <dmitry.gladkov@intel.com>
Signed-off-by: Sean Hefty <sean.hefty@intel.com>

jswaro commented on include/ofi_mr.h:
```diff
@@ -0,0 +1,123 @@
+{
+	int ret;
+
+	ret = eq->mr_monitor->subscribe(eq->mr_monitor, addr, len, subscript);
```
Unless the underlying implementation is responsible for initializing the subscription, the subscript is returned unmodified. If the subscript was passed in zero'd out, it crashes at the point of trying to retrieve events from the monitor.

jswaro replied:
Also, is it assumed the cache will set the 'events' field?

gladkovdmitry17 replied:
> subscribe->eq = eq;

Thanks, it should be there

jswaro commented on prov/util/src/util_mr_monitor.c:
```diff
@@ -0,0 +1,115 @@
+	fastlock_destroy(&eq->lock);
+}
+
+static void util_mr_monitor_read_events(struct ofi_mr_monitor_eq *eq)
```
I think this was commented on in the last PR. Is this going to stay as as 'util_mr_monitor_*' instead of 'ofi_mr_monitor'?

gladkovdmitry17 replied:
The Utility code uses the following style as I noticed:
- external functions (like API) are declared with `ofi_[utility_module]_*` prefix
- internal functions (that is used in this file only) are declared with `util_[utility_module]_*` prefix

For example, looking at the CQ utility code (util_cq.c):
`ssize_t ofi_cq_read(struct fid_cq *cq_fid, void *buf, size_t count)`
`static void util_cq_read_tagged(void **dst, void *src)`

jswaro commented on include/ofi_mr.h:
```diff
@@ -0,0 +1,123 @@
+}
+
+static inline
+void ofi_mr_monitor_unsubscribe(struct ofi_mr_subscription *subscript)
```
nit: can we just expand the name from 'subscript' to 'subscription'? It's three more letters.

jswaro commented on include/ofi_mr.h:
```diff
@@ -0,0 +1,123 @@
+void ofi_mr_monitor_finalize(struct ofi_mr_monitor *mr_monitor);
+void ofi_mr_monitor_eq_add(struct ofi_mr_monitor *mr_monitor,
+			   struct ofi_mr_monitor_eq *eq,
+			   int events);
```
Do mr subscriptions inherit the events flag from the eq? If so, it would explain my question from below.

gladkovdmitry17 replied:
I've designed it to work as follows:
User of MR monitor calls `ofi_mr_monitor_eq_add` and passes the events that user would expect from the `ofi_mr_monitor_get_event` call.

Answering on you question below:
> Also, is it assumed the cache will set the 'events' field?

the events flag should be set by ab underlying notifier when an event is generated.
i.e. if user request MAPPED and UNMAPPED event flags for the EQ, underlying notifier should set an appropriate event flag to a `ofi_mr_subscription` entry.

jswaro replied:
> User of MR monitor calls ofi_mr_monitor_eq_add and passes the events that user would expect from the ofi_mr_monitor_get_event call.

That follows what I expected -- I want to confirm that it should be the responsibility of the underlying notifier to check the EQ event flags and discard generated events that the EQ hasn't requested. 

IE. if the underlying notifier generates MAPPED event, but the EQ only has the UNMAPPED enum value set, then the underlying implementation discards the event. 

If not, who is responsible for filtering the events?

shefty replied:
The underlying provider filters the events, but we can easily push this up to handle the case where the provider always gets all events

jswaro commented:
How is memory management for the events being handled?  

The subscription generated with 'ofi_mr_monitor_eq_add' cannot be used for a variety of reasons. To me, this implies the underlying notifier allocates the memory for the event generated, but there appears to be no way for it to know that the upper layer is done with it.

shefty commented:
eq_ad only associates an event queue for demultiplexing purposes.  The subscriptions are the events reporting mechanism, and are allocated by the user.

shefty commented:
Updated - Fixed subscription initialization.  Added reference counting on the eq.  Fixed get_event handler callback.  Added asserts to validate subscription event reporting.

jswaro commented:
> The subscription generated with 'ofi_mr_monitor_eq_add' cannot be used for a variety of reasons
> > The subscriptions are the events reporting mechanism, and are allocated by the user.

Ok, so the subscription is provided by the upper layer, that's fine. However, that doesn't appear to resolve the problem that I attempted to pose. 

Once there are multiple types of events to report, and the subscription is persistent (as it is with this model), you can receive multiple events between individual calls to the monitor for the same subscription. Given the discussion from #3548, that multiple events could be returned, then how are multiple events resolved through this single subscription that also seems to be carrying event data? 

I have a test application, that could easily extended to demonstrate the case. How is it resolved? What are the expectations? I'd really like to see them defined before this is committed.

shefty commented:
This proposal does not indicate what event occurred, only that some event occurred.  The case of multiple events occurring is handled by the proposed code.  If we want to report what event occurred, along with the order that those events have occurred, then we're either looking at dynamically allocating structures or using some sort of fixed pool/queue (which can result in event loss).

jswaro commented:
Ok, I can see that. What is this proposal then? I think this is a discussion that could be taken offline into a call. 

From the original code, I don't see a clear mapping from that code, to this code -- based on my current understanding of intents. 

> If we want to report what event occurred, along with the order that those events have occurred, then we're either looking at dynamically allocating structures or using some sort of fixed pool/queue (which can result in event loss).

If the type of event is not reported, how does the upper layer know what to do with the event?

shefty commented:
I don't understand your comment.  The code handles a subscription receiving multiple events between subsequent calls to get_events().  It does not report what events occurred, how many, the pages that were affected, the final event state of the region, etc.  Reporting any of that likely requires significantly more complexity, that I honestly don't even think is necessary.  Just knowing that there's been some change to the address to page mappings is likely sufficient for the user.  Besides, my understanding of the gni 'monitor' is that it can only report one type of event anyway.  Added to that by the time an event is being processed, additional events could have been generated and queued, which makes fine grained event reporting of questionable use.

jswaro commented:
> It does not report what events occurred, how many, the pages that were affected, the final event state of the region, etc.

This is relevant to the idea of a MAPPED event. As an application writer, I may choose to use scalable registration, and potentially -- if the hardware allows it -- register my entire address space. If the application touches some of those pages, and they then become mapped -- the provider will need to know what pages were mapped so that the registration can become updated. It may be an invalid assumption to believe the hardware will accept the entire range of the registration, and attempt to figure out what pages were faulted/paged in. 

As far as UNMAPPED is concerned, you are right -- it doesn't matter much. The registration is invalid, and cannot/should not be used. This is the simple case. 

However, without some context as to know what has happened to the region, how does the cache know what to do with the registration? In the original design, that wasn't a problem because there was only one type of event to support, with a clear action for the cache to take with respect to that. 

> Reporting any of that likely requires significantly more complexity, that I honestly don't even think is necessary.

Again, this goes into the assumption that the hardware or software surrounding the hardware is capable of updating the registration without some knowledge of what in memory was just changed. The GNI provider, and more specifically, uGNI does not support updating a registration over the entire address range of a registration if some part of that address range is unmapped. If it occurs for one provider, it surely could occur with another. 

> Just knowing that there's been some change to the address to page mappings is likely sufficient for the user.

The purpose of this work is to hide it from the user so that the application isn't required to do all of the interaction that is being required with mode flags such as FI_MR_MMU_NOTIFY. The provider may need to know what range was affected to notify the hardware that range X has changed and the registration should be updated to reflect that. 

> Added to that by the time an event is being processed, additional events could have been generated and queued, which makes fine grained event reporting of questionable use.

I agree. 

> Besides, my understanding of the gni 'monitor' is that it can only report one type of event anyway.

That is a byproduct of using KDREG as the monitoring interface. Another byproduct of using KDREG is the fact that it would only generate one event per subscription, because once the region was unmapped in some way, hardware and software would have two very different perspectives of memory mappings -- as such, until it is re-registered, there is no reason to monitor the region. Since there was no way to reconcile when a registration was remapped, this simplified the design. 

With the abstraction from the GNI provider code to common core that we are working on here, some of those problems are coming forward with changes to the design. What I'm trying to do is illustrate the issues so that they can be addressed.

gladkovdmitry17 commented:
@shefty @jswaro 
Just to summarize, the MAPPED event won't be used for the MR cache. it can be exposed by providers in the future if needed.

> As an application writer, I may choose to use scalable registration, and potentially -- if the hardware allows it -- register my entire address space. If the application touches some of those pages, and they then become mapped -- the provider will need to know what pages were mapped so that the registration can become updated.

I agree, but  the MR monitor doesn't assume that the memory is registered by fi_mr_reg t all. It merely provides a mechanism to monitor memory that is allocated/freed by a process. And these facts can be used by provider/apps to take some actions (please, note if we're saying about MR, seems only MAPPED event is acceptable and should be used).
Let's drop MR from name to avoid any confusions. I guess that name ofi_mem_monitor is good enough for this purpose.

For example we could overwrite sbrk() call (as it is done in the https://github.com/adobe/chromium) and receive events about new registered regions. But to provide this, we need to extend ofi_mr_subscription (or ofi_mem_subcription) by addr, length fields that is being allocated. 
My idea is to prepare common framework that provides Monitoring of all events. But for now it will use only UNMAPPED event for MR caching needs.

If you're agree I could extend APIs and structures tomorrow to be able correctly accept MAPPED events. And I would rename this to avoid further misunderstanding of that due to the incorrect name.

shefty commented:
I'll update the patch to rename the function and remove 'future extensions'.  We need something that a.) we can actually implement, and b.) is usable.  Designing around theoretical extensions that aren't immediately implemented or used is a waste of time, and should be done when we have an actual user of the feature.

@gladkovdmitry17 - Does a monitor that simply indicates that there has been 'some' change to the virtual memory mappings work for the cache?

shefty commented:
Patch updated:

Renamed mr_monitor to mem_monitor.  The event queue has been renamed to notification queue. The API has been cleaned up -- len converted from uint64_t to size_t, removed 'event' capability flags, added addr + len parameters to unsubscribe call, reference counting replaces lists.

gladkovdmitry17 commented:
> Does a monitor that simply indicates that there has been 'some' change to the virtual memory mappings work for the cache?

@shefty Yes, this works for the cache. MR cache is interested in UNMAPPED events and if the memory NQ will notify about some events, then MR cache will consider this event as UNMAPPED

>Patch updated
Thanks!

gladkovdmitry17 commented on Makefile.am:
```diff
@@ -36,30 +38,31 @@ rdmaincludedir = $(includedir)/rdma
+	prov/util/src/util_mr.c		\
+	prov/util/src/util_ns.c		\
+	prov/util/src/util_shm.c	\
+	prov/util/src/util_mr_monitor.c
```
Let's rename this file to either `util_monitor.c` or `util_mem_monitor.c`.
I prefer the second one. Seems this name reflects the purpose of this functionality in more detail

jswaro replied:
I agree with that. The header file should probably be renamed as well.

shefty commented:
Updated - rename .c file to util_mem_monitor.c.  Kept name ofi_mr.h, as we'll eventually move all memory region/mgmt defines into this file.  Will merge after CI completes.

gladkovdmitry17 commented:
@shefty I've fixed build for clang/LINUX and added needed stuff for Windows.

The all changes are in #3581. Can we switch to that PR? Or I can wait until this PR is merged and I re-base the #3581 PR on top of these changes. This also works for me :)

shefty commented:
Replaced by PR #3581.
